### ISSUE TITLE
Undoing of normalization by world size?

### ISSUE THREAD

**Issue by emjotde** [pytorch/fairseq#494]
Hi,
I am rummaging a bit through your code and am curious about this line:

https://github.com/pytorch/fairseq/blob/master/fairseq/trainer.py#L245

```
# normalize grads by sample size
self.optimizer.multiply_grads(self.args.distributed_world_size / float(sample_size))
```

Since you multiply with `distributed_world_size` are you somehow undoing an average over `distributed_world_size` that happened elsewhere? If yes, where would that happen?

**Comment by myleott** [comment]
Yes, exactly. DistributedDataParallel averages over the number of workers: https://pytorch.org/docs/master/nn.html#torch.nn.parallel.DistributedDataParallel

**Comment by emjotde** [comment]
Gosh, I stared at that page for a while before I asked and then some more after you pointed me to it. This really should be one of those big fat red warnings :)
Thanks a lot.
